pallets/parachain-staking/src/lib.rs

```diff
@@ -2773,95 +2383,121 @@ pub mod pallet {
+				..
+			} = <TotalCollatorStake<T>>::get();
+			let staking_rate = Perquintill::from_rational(total_collators, total_issuance);
+
+			InflationConfig::<T>::get()
+				.collator
+				.compute_reward::<T>(stake, staking_rate, multiplier)
+		}
+
+		/// Calculates the delegator staking rewards for `multiplier` many
+		/// blocks based on the given stake.
+		///
+		/// Depends on the current total issuance and staking reward
+		/// configuration for delegators.
+		fn calc_block_rewards_delegator(stake: BalanceOf<T>, multiplier: BalanceOf<T>) -> BalanceOf<T> {
+			let total_issuance = T::Currency::total_issuance();
+			let TotalStake {
+				delegators: total_delegators,
+				..
+			} = <TotalCollatorStake<T>>::get();
+			let staking_rate = Perquintill::from_rational(total_delegators, total_issuance);
+
+			InflationConfig::<T>::get()
+				.delegator
+				.compute_reward::<T>(stake, staking_rate, multiplier)
+		}
+
+		/// Increment the accumulated rewards of a collator by consuming their
+		/// current rewards counter.
+		///
+		/// Moreover, automatically increments all associated delegators since
+		/// their reward increment is dependend on the collator reward counter.
+		///
+		/// Resets all reward counters of the collator and their delegators to
+		/// zero.
+		fn do_inc_collator_reward(collator: &T::AccountId, stake: BalanceOf<T>) -> usize {
+			// get reward counters
+			let col_reward_count = RewardCount::<T>::get(collator);
+
+			// set reward data for collator
+			Rewards::<T>::mutate(collator, |reward| {
+				*reward = reward.saturating_add(Self::calc_block_rewards_collator(stake, col_reward_count.into()));
+			});
+
+			// set reward data for delegators
+			let num_delegators = if let Some(state) = CandidatePool::<T>::get(collator.clone()) {
+				let num_delegators = state.delegators.len();
+				for Stake { owner, amount } in state.delegators {
+					Self::do_inc_delegator_reward(&owner, amount, collator);
+					// Reset delegator counter since collator counter will be reset
+					RewardCount::<T>::insert(owner, 0);
+				}
+				num_delegators
+			} else {
+				0usize
+			};
+
+			// Reset collator reward count
+			RewardCount::<T>::insert(collator, 0);
+
+			num_delegators
+		}
+
+		/// Increment the accumulated rewards of a delegator by consuming their
+		/// current rewards counter. The counter will be reset to the collator
+		/// counter.
+		fn do_inc_delegator_reward(acc: &T::AccountId, stake: BalanceOf<T>, col: &T::AccountId) -> Weight {
+			// get reward counters
+			let del_reward_count = RewardCount::<T>::get(acc);
+			let col_reward_count = RewardCount::<T>::get(col);
+			let diff = col_reward_count.saturating_sub(del_reward_count);
+
+			// only update if collator has higher reward count
+			if diff > 0 {
+				Rewards::<T>::mutate(acc, |reward| {
+					*reward = reward.saturating_add(Self::calc_block_rewards_delegator(stake, diff.into()));
+				});
+				RewardCount::<T>::insert(acc, col_reward_count);
```

Comment:
When you request a payout as a delegator, the number of rewards you receive is dependent on your counter and the one of your collator. A delegator's counter can at most be even with the corresponding collator counter. When a delegator increments their rewards, the counter is equalized is aligned with the collator one to reflect that the rewards can only be incremented further after the collator authored at least another block.

_Example: 
1. Collator Alice has counter 100 and Bob joins as a delegator. Since he should only receive rewards from now on, his counter is initialized with 100. 
2. When Alice's counter is at 200, Bob stakes more and triggers a reward increment. He receives `(200-100)*reward_per_block(Bob)`. He is then treated as a newcomer, thus "resetting" his counter to 200._

Reply:
I will add a comment to explain why it is not reset to zero.

Reply:
Ok, understood. Thanks.

Reply:
I guess it's just a design decision, whether to substract the collator count from the total, or sum the delegator count to the collator count. I would have gone the second road as it makes it easier to understand, but yeah a comment explaining that would help.

Reply:
[ad13650](https://github.com/KILTprotocol/kilt-node/pull/384/commits/ad13650e92a6c87335b1ac3374bfc03531c6025f)

Reply:
How would your proposal work? We basically don't want to increase the delegator counter ever, just when initializing/resetting, we need to align with the collator one in my case. I decided for that to decrease storage writes.